cmds/webboot/webboot.go

```diff
@@ -28,21 +34,65 @@ func (i *ISO) exec(uiEvents <-chan ui.Event, boot bool) error {
 		return err
 	}
 	verbose("Get configs: %+v", configs)
-	if boot {
-		entries := []menu.Entry{}
-		for _, config := range configs {
-			entries = append(entries, &Config{label: config.Label()})
+	if !boot {
+		return nil
+	}
+	entries := []menu.Entry{}
+	for _, config := range configs {
+		entries = append(entries, &Config{label: config.Label()})
+	}
+	c, err := menu.DisplayMenu("Configs", "Choose an option", entries, uiEvents)
+	if err == nil {
+		err = bootiso.BootFromPmem(i.path, c.Label())
+	}
+	return err
+}
+
+// getCachedDirectory recognizes the usb stick that contains the cached directory from block devices,
+// and return it's mount point.
+// the file structure of USB stick could be
+// +-- USB root
+// |  +-- Images (<--- the cache directory. It must be named as "Images")
+// |     +-- subdirectories or iso files
+// ...
+func getCachedDirectory() (*mount.MountPoint, error) {
+	blockDevs, err := block.GetBlockDevices()
+	if err != nil {
+		return nil, fmt.Errorf("No available block devices to boot from")
+	}
+
+	mountPoints, err := ioutil.TempDir("", "temp-device-")
+	if err != nil {
+		return nil, fmt.Errorf("Cannot create tmpdir: %v", err)
+	}
+
+	for _, device := range blockDevs {
+		mp, err := mount.TryMount(filepath.Join("/dev/", device.Name), filepath.Join(mountPoints, device.Name), "", mount.ReadOnly)
+		if err != nil {
+			continue
 		}
-		if c, err := menu.DisplayMenu("Configs", "Choose an option", entries, uiEvents); err == nil {
-			if err := bootiso.BootFromPmem(i.path, c.Label()); err != nil {
-				return err
-			}
-			return nil
-		} else {
-			return err
+		if _, err = os.Stat(filepath.Join(mp.Path, "Images")); err == nil {
+			return mp, nil
 		}
 	}
-	return nil
+	return nil, fmt.Errorf("Do not find the cache directory: Expected a /Images under at the root of a block device(USB)")
+}
+
+// UseCacheOption's exec displays all possible cache directory
+// if the -dir args is valid show it in the menu
+// if it's able to find a cached diretory in the USB stick, show it too
+func (u *UseCacheOption) exec(uiEvents <-chan ui.Event, dir string) (menu.Entry, error) {
+	// filefath.Join an empty string to make sure the format of path is consistent
+	cacheDir = filepath.Join(dir, "")
```

Comment:
filepath.Clean may be more useful here.
https://play.golang.org/p/IHPw1QDR85l
"Join joins any number of path elements into a single path, separating them with an OS specific Separator. Empty elements are ignored. The result is Cleaned. "

in other words Join calls Clean.

Reply:
Thank you! That's helpful!

Reply:
Ron I find there is a different between filepath.Clean and filepg.Join empty element, that is , when the raw path is empty too, the Clean will returns ".", while the Join will returns ""  Just for let you know.

Reply:
What is the right one to have? Which one is preferred?

Reply:
I think the clean one is what the filepath want me to use in this situation, but for here I do not want it return "." when the dir is empty. Currently I still use clean, but I move the clean sentence inside to the if statement , so the if statement have checked that the dir will not be empty.